### PR TITLE
Part 1 of RFC2906 - Packages can inherit fields from their root workspace

### PR DESCRIPTION
**Pull request by Muscraft**
Tracking issue: #8415
RFC: rust-lang/rfcs#2906

All changes were heavily inspired by #8664 and #9684

A big thanks to both @yaymukund and @ParkMyCar. I pulled a lot of inspiration from their branches.

I would also like to thank @alexcrichton for all the feedback on the first attempt. It has brought this into a much better state.

All changes have been made according to the RFC as well as @alexcrichton's [comment](https://github.com/rust-lang/cargo/pull/8664#issuecomment-704878103). 

This is part 1 of many, as described by [this comment](https://github.com/rust-lang/cargo/pull/9684#issuecomment-943567692), [this comment](https://github.com/rust-lang/cargo/pull/9684#pullrequestreview-779070283) and redefined  [by this one](https://github.com/rust-lang/cargo/pull/10497#issuecomment-1076301312).

This PR focuses on inheriting in root package, including:
- Add `MaybeWorkspace<T>` to allow for `{ workspace = true }`
- Add a new variant to `TomlDependency` to allow inheriting dependencies from a workspace
- Add `InheritableFields` so that we have somewhere to get a value from when we `resolve` a `MaybeWorkspace<T>`
  - `license_file` and `readme` are in `InheritableFields` in this part but are not `MaybeWorkspace` for reasons [described here](https://github.com/rust-lang/cargo/pull/10497#issuecomment-1076470424)
- Add a method to `resolve` a `MaybeWorkspace<T>` into a `T` that can fail if we have nowhere to pull from or the workspace does not define that field
- Disallow inheriting from a different `Cargo.toml`
  - This means that you can only inherit from inside the same `Cargo.toml`, i.e. it has both a `[workspace]` and a `[package]`
  - Forcing this requirement allows us to test the implementations of `MaybeWorkspace<T>` and the new `TomlDependency` variant without having to add searching for a workspace root and the complexity with it

Remaining implementation work for the RFC
- Support inheriting in any workspace member
- Correctly inherit fields that are relative paths
  - Including adding support for inheriting `license_file`,  `readme`, and path-dependencies
-  Path dependencies infer version directive
- Lock workspace dependencies and warn when unused
- Optimizations, as needed
- Evaluate any new fields for being inheritable (e.g. `rust-version`)
- Evaluate making users of `{ workspace = true }` define the workspace they pull from in `[package]`

Areas of concern:
- `TomlDependency` Deserialization is a mess, and I could not figure out how to do it in a cleaner way without significant headaches. I ended up having to do the same thing as last time [which was an issue](https://github.com/rust-lang/cargo/pull/9684#discussion_r728444103).
- Resolving on a `MaybeWorkspace` feels extremely verbose currently:
  ```rust
  project.homepage.clone().map(|mw| mw.resolve(&features, "homepage", || get_ws(inheritable)?.homepage())).transpose()?,
  ```
  This way allows for lazy resolution of inheritable fields and finding a workspace (in part 2) so you do not pay a penalty for not using this feature. The other bit of good news is `&features` will go away as it is just feature checking currently.

- This feature requires some level of duplication of code as well as remaking the original `TomlManifest` which adds extra length to the changes.
- This feature also takes a linear process and makes it potentially non-linear which adds lots of complexity (which will get worse in Part 2)

Please let me know if you have any feedback or suggestions!

### CODE REVIEW

**Comment by epage** [comment]
> This is part 1 of 2 - 3 as described by https://github.com/rust-lang/cargo/pull/9684#issuecomment-943567692 as well as https://github.com/rust-lang/cargo/pull/9684#pullrequestreview-779070283.

As a heads up the "miscellaneous" aspect of part 3 is about to grow.  `cargo add` is in FCP, see https://github.com/rust-lang/cargo/pull/10472.

I think it could make sense to break this out into a step 4.  I could help out of implement that part.  I've recorded [more details on the tracking issue](https://github.com/rust-lang/cargo/issues/8415#issuecomment-1075544790).

File: src/cargo/util/toml/mod.rs
Comment (epage) [inline src/cargo/util/toml/mod.rs:1008]: Might be good to say tell the user why.

Also, errors generally start with lowercase

File: src/cargo/util/toml/mod.rs
Comment (epage) [inline src/cargo/util/toml/mod.rs:1047]: How do we decide on handling fields not specified in the RFC?

The RFC doesn't mention `rust_version` but it was [stablized on Aug 2nd](https://github.com/rust-lang/cargo/pull/9732) which is after this RFC was written.

File: src/cargo/util/toml/mod.rs
Comment (epage) [inline src/cargo/util/toml/mod.rs:1001]: Why do we need both instead of just `impl FnOnce() -> Option<T>`?

If we switch it to `impl FnOnce( -> CargoResult<T>` then could we get rid of the `label` parameter?  I guess it would make the error less nice for the `workspace = false` case.  Unsure on the trade off there.

File: src/cargo/util/toml/mod.rs
Comment (epage) [inline src/cargo/util/toml/mod.rs:1366]: I feel like `get_ws` should either be renamed here or in `resolve`.  We are using a `get_ws` function for the `inherit` parameter and then we have a `get_ws` parameter.

File: src/cargo/util/toml/mod.rs
Comment (epage) [inline src/cargo/util/toml/mod.rs:1664]: Am I reading this correctly that this is equivalent of
```rust
value
    .map(|mw| mw.resolve(&features, label, || get_ws(inheritable_fields), resolve))
    .transpose()
```

File: src/cargo/util/toml/mod.rs
Comment (epage) [inline src/cargo/util/toml/mod.rs:1711]: Is `i.license_file)()` already an absolute path by this point?

File: src/cargo/util/toml/mod.rs
Comment (epage) [inline src/cargo/util/toml/mod.rs:2249]: ditto about this error message

File: src/cargo/util/toml/mod.rs
Comment (epage) [inline src/cargo/util/toml/mod.rs:2237]: ditto about API comments

File: src/cargo/util/toml/mod.rs
Comment (epage) [inline src/cargo/util/toml/mod.rs:2263]: Are paths resolve at this point?

File: src/cargo/util/toml/mod.rs
Comment (epage) [inline src/cargo/util/toml/mod.rs:2138]: It looks like `readme.workspace = true` is causing the default readme to be loaded rather than the workspaces

**Review by epage — COMMENTED** [review]
Thanks for doing this! I'm very excited by this feature!

I'm taking a break for now; all I have left are the tests.

Could you be more explicit in what the parts are?  I might have missed it but looking at the past comments, I'm not seeing "Path dependencies infer version directive" mentioned and  wanting to make sure that we are taking that into account for closing out the RFC.

File: src/cargo/util/toml/mod.rs
Comment (Muscraft) [reply src/cargo/util/toml/mod.rs:1008]: would "tried to inherit from a workspace for {}, but the workspace is not allowed to specify inheritable fields at this time" be better?

File: src/cargo/util/toml/mod.rs
Comment (Muscraft) [reply src/cargo/util/toml/mod.rs:1664]: You are reading it correctly and I completely forgot about 
```rust
 .transpose()
```
I think that fits much better than a dedicated function so I will change them all to that!

File: src/cargo/util/toml/mod.rs
Comment (Muscraft) [reply src/cargo/util/toml/mod.rs:1366]: I agree with you on this, I think we should rename it in `resolve` but I will play around with what feels best

File: src/cargo/util/toml/mod.rs
Comment (Muscraft) [reply src/cargo/util/toml/mod.rs:1711]: I don't think it's an absolute path by this point. I think it technically needs special handling to make this a path as we resolve it, but that is all should be done in part 2 or 3

File: src/cargo/util/toml/mod.rs
Comment (Muscraft) [reply src/cargo/util/toml/mod.rs:2263]: No, paths aren't resolved at all in part 1. This only works here since we can only inherit from a `Cargo.toml` that has both a `[workspace]` and a `[package]`

File: src/cargo/util/toml/mod.rs
Comment (Muscraft) [reply src/cargo/util/toml/mod.rs:2138]: You are correct. I will make this work correctly

File: src/cargo/util/toml/mod.rs
Comment (epage) [reply src/cargo/util/toml/mod.rs:1711]: Could you note this explicitly in the PR description so its easier to bring others up to speed on whats included and whats not?

File: src/cargo/util/toml/mod.rs
Comment (epage) [reply src/cargo/util/toml/mod.rs:1008]: Under what condition do we not have inheritable fields?  I'm looking at `deny_inherit_fields_from_parent_workspace` and `deny_inherit_dependencies_from_parent_workspace` and its not too clear to me.

Comment (Muscraft) [reply src/cargo/util/toml/mod.rs:1008]: I'll amend my post above to better explain this but ill also describe it here. 

Part 1 has no tooling to allow inheriting from a workspace that is in a different `Cargo.toml`. This mean that you can only inherit from inside the same `Cargo.toml`, i.e. it has both a `[workspace]` and a `[package]`. This was done so that we could split the implementation of `Maybeworkspace<T>` and `Workspace(TomlWorkspaceDependency)` up from the searching for a workspace root and keeping the cache and various other things. By allowing inheriting from the same `Cargo.toml` only, it makes these changes testable without adding in the searching and other various things in the same that PR caused too much complexity in the previous PR.

File: src/cargo/util/toml/mod.rs
Comment (Muscraft) [reply src/cargo/util/toml/mod.rs:1047]: I think most things in `TomlProject` can be added to this but I would have to check the specifics for each one. As for `rust_version`, we should be able to add it, but we should explicitly state we did and potentially modify the RFC to reflect this if possible. I will happily add it if requested since it should only be a few extra lines if I'm looking at this correctly.

File: src/cargo/util/toml/mod.rs
Comment (Muscraft) [reply src/cargo/util/toml/mod.rs:1001]: The [original comment](https://github.com/rust-lang/cargo/pull/9684#discussion_r728453774) had a `label` and was `get_ws: impl FnOnce() -> &T`. I modified it to what it is now for convenience but I think no matter what we should leave `label` since it makes the error messages nicer in the current implementation.

As for going to `impl FnOnce() -> Option<T>`, I will look into this and report back. If I remember to when I changed it to this, it made things a bit easier to understand and split getting the fields to inherit from and how to get it from those inheritable fields. I can see why one might be a better implementation since that way if we already have inheritable fields we could directly call in without `inherit: impl FnOnce() -> Option<&'a InheritableFields>,`

File: src/cargo/util/toml/mod.rs
Comment (Muscraft) [reply src/cargo/util/toml/mod.rs:1711]: will do!

**Comment by Muscraft** [comment]
> Could you be more explicit in what the parts are? I might have missed it but looking at the past comments, I'm not seeing "Path dependencies infer version directive" mentioned and wanting to make sure that we are taking that into account for closing out the RFC.

I edited my original comment/explanation to better explain what is happening in each part. Let me know if I need to change anything or make the parts clearer,

File: src/cargo/util/toml/mod.rs
Comment (epage) [reply src/cargo/util/toml/mod.rs:1008]: So it sounds like this error is an internal error about the feature not being implemented yet?  Seems like that is what we should have the error message focus on.  Something like "Inheriting from a parent workspace is not implemented yet".

I see how your original message and the follow up were intending to say that but there was just enough wiggle room for misinterpretation that I didn't pick this up.

Could you also clarify that with the test names?  Something like "error_on_unimplemented_inheritance()`.  This will make it clearer that the contract the test is enforcing is temporary rather than being fundamental.

File: src/cargo/util/toml/mod.rs
Comment (epage) [reply src/cargo/util/toml/mod.rs:1001]: >  but I think no matter what we should leave label since it makes the error messages nicer in the current implementation.

This is why I was wondering about the signature `impl FnOnce( -> CargoResult<T> ` as it could do the error formatting as well.  Granted, that makes it so we duplicate the error messages which makes it more likely for them to be unintentionally different, so maybe keeping it an `Option` with `label` is better.

**Comment by epage** [comment]
> I edited my original comment/explanation to better explain what is happening in each part. Let me know if I need to change anything or make the parts clearer,

Thanks!  Its helpful when reviewing to have the full context of what is expected to be in scope and out of scope so I know what to look for.

I'm thinking it would be good to break up Part 2 to be more focused so the review process is easier for everyone.

My suggestion for different PRs

- Support for inheriting in root package (this PR)
- Support inheriting in any workspace member
- Correctly inherit fields that are relative paths
  - I could see delaying support for inheriting these fields until this PR but I can go either way
- Path dependencies infer version directive
- Optimizations, as needed (PR per optimization)

These are
- User-facing / testable
- Narrow scoped
- Single purpose
- Mostly parallizable

When PRs try to fulfill too many purposes
- It can be unclear which purpose a changed line is helping, requiring more time from the reviewer
- Things are more likely to be missed in each pass but then found in a later one, leading to a lot of back-and-forth which can be frustrating and discouraging for everyone
- A single point of concern can hold up every improvement
- The longer and more iterations a PR goes through, the more conflicts that will have to be handled by the author.

File: src/cargo/util/toml/mod.rs
Comment (Muscraft) [reply src/cargo/util/toml/mod.rs:1008]: I'll change it to what you said. This could happen in Part 2 if you can't find a workspace root but I think we could just change the message in Part 2

**Comment by Muscraft** [comment]
> My suggestion for different PRs
> - Support for inheriting in root package (this PR)
> - Support inheriting in any workspace member
> - Correctly inherit fields that are relative paths
>   - I could see delaying support for inheriting these fields until this PR but I can go either way
> - Path dependencies infer version directive
> - Optimizations, as needed (PR per optimization)

Support inheriting in any workspace member, and Correctly inherit fields that are relative paths to me are fairly tied together since this would delay `license-file`, `readme`, and `dependencies` (I think this is all that could be relative paths) until part 3. It would require setting up tests in Part 2 to disallow these things and error messages and testing for them could be a bit odd. 

If we were to split these into Part 2 and Part 3, I think Part 1 should remove `MaybeWorkspace<T>` from `license-file` and `readme` and only add them back in Part 3. We can leave `dependencies` the same as we have now since its `resolve` is separate and could be updated in Part 3/4.

I'm flexible on how we want to go about this and will happily do whatever is easiest to review.

**Comment by epage** [comment]
> If we were to split these into Part 2 and Part 3, I think Part 1 should remove MaybeWorkspace<T> from license-file and readme and only add them back in Part 3. We can leave dependencies the same as we have now since its resolve is separate and could be updated in Part 3/4.

As I said, I'm fine with it if you want to remove them from this PR. 

> I'm flexible on how we want to go about this and will happily do whatever is easiest to review.

Let's move forward with this then so this has the best chances of making it in!

**Comment by Muscraft** [comment]
Let me know your thoughts on the newest commit, I believe I made all the changes you had mentioned so far.

Changing to a `impl FnOnce() -> CargoResult<T>` did have some benefits and made things less cluttered overall, but it made the getters for `InheritableFields` a  little verbose. If the getters need to be fixed, we could do a `macro_rules!` but I dislike this for readability and maintenance reasons. If you have any suggestions I think it might be a pain point.

What are your thoughts on adding `rust_version` [as mentioned here](https://github.com/rust-lang/cargo/pull/10497#discussion_r832584924)? Personally, I think we could add it if not, adding it to a part down the line is perfectly fine as well. It also could be added with a separate PR not part of this RFC since it was not mentioned in the RFC. 

I will update what is being done in the parts in the main part of my PR so others can see what is going on easier.

File: src/cargo/core/workspace.rs
Comment (epage) [inline src/cargo/core/workspace.rs:1740]: Using `[]` makes sense for `dependencies` but not as much for regular values like these.  Maybe we should consistently just do back-ticks for all of them including `dependencies`?

File: src/cargo/util/toml/mod.rs
Comment (epage) [inline src/cargo/util/toml/mod.rs:2205]: Please re-phrase the comment to be a parameter to the `unreachable!` macro, it accepts a value so we might as well use it and tell the user whats going on rather than having to dig in to find the comment for what invariant was violated.

**Comment by epage** [comment]
> Changing to a `impl FnOnce() -> CargoResult<T>` did have some benefits and made things less cluttered overall, but it made the getters for `InheritableFields` a little verbose. If the getters need to be fixed, we could do a `macro_rules!` but I dislike this for readability and maintenance reasons. If you have any suggestions I think it might be a pain point.

I think it looks great.  Putting it on the inheritable fields helps keep all the error reporting in one place so we can keep them in sync.
 
> What are your thoughts on adding `rust_version` [as mentioned here](https://github.com/rust-lang/cargo/pull/10497#discussion_r832584924)? Personally, I think we could add it if not, adding it to a part down the line is perfectly fine as well. It also could be added with a separate PR not part of this RFC since it was not mentioned in the RFC.

I think this is a wider conversation that should be addressed independent of this PR.  I've gone ahead and updated the tracking issue with a check box for this.

File: tests/testsuite/deduplicate_workspace.rs
Comment (epage) [inline tests/testsuite/deduplicate_workspace.rs:1]: nit: Should we update this comment and filename to refer to "inheritable" rather than "de-duplicate" as that is how we are referring to it everywhere else in this PR (documentation, feature flag, types, etc)

File: tests/testsuite/deduplicate_workspace.rs
Comment (epage) [inline tests/testsuite/deduplicate_workspace.rs:130]: Nit: with TOML now supporting dotted keys, should we stylistically be preferring `version.workspace = true`?

Doesn't need to be addressed now but what tone we set in rolling out this feature will likely be copied by people going forward.

File: tests/testsuite/deduplicate_workspace.rs
Comment (epage) [inline tests/testsuite/deduplicate_workspace.rs:498]: I think I'd do
```
error inheriting `description` from workspace root manifest's `workspace.description`
```
- the more specific action "inherit" over "read" which also differentiates it more from the "Caused by"
- Just using back-ticks without `[]`

File: tests/testsuite/deduplicate_workspace.rs
Comment (epage) [inline tests/testsuite/deduplicate_workspace.rs:501]: I think I'd change this to
```
`workspace.description` was not defined
```
If we could say in what file its not defined, that'd be even better.  It should be obvious but it doesn't help.

File: tests/testsuite/deduplicate_workspace.rs
Comment (epage) [inline tests/testsuite/deduplicate_workspace.rs:727]: I think I'd do
```
`workspace=false` is unsupported for  `package.description`
```

**Review by epage — COMMENTED** [review]
Finally got to looking at the tests

File: src/cargo/core/workspace.rs
Comment (Muscraft) [reply src/cargo/core/workspace.rs:1740]: I will do whatever works best, I think I only used `[]` due to past attempts using it/seeing it before. If you think using backticks and making it \`workspace.version\` is better across the board we will do that.

**Comment by epage** [comment]
@Muscraft I hope you don't mind, I made some edits to the PR description to try make the intent clearer to bring others on board more quickly as I share this.

I also updated the relative path handling to include path-dependencies on top of `readme` and `license_file`, see https://github.com/rust-lang/rfcs/blob/master/text/2906-cargo-workspace-deduplicate.md#effect-resolution-for-relative-path-dependencies

Looking over the RFC, a couple more items I've noticed that we don't have covered
- Should we have a test to verify `cargo metadata` and `cargo read-manifest`s behavior?
- Looks like we should add  to our task list the `Cargo.lock` handling as I don't think I saw it in this PR

File: src/cargo/util/toml/mod.rs
Comment (Muscraft) [reply src/cargo/util/toml/mod.rs:2205]: I updated this to be
```rust
TomlDependency::Workspace(_) => {
    unreachable!(
        "We check that no workspace defines dependencies with \
        `{{ workspace = true }}` when we read a manifest from a string. \
        this should not happen but did on {}",
        label
    )
},
```

I think it gives a better description of what is going on and where we check to make sure this never happens.

File: tests/testsuite/deduplicate_workspace.rs
Comment (Muscraft) [reply tests/testsuite/deduplicate_workspace.rs:1]: do you like `//! Tests for inheriting Cargo.toml fields with { workspace = true }` and `inheritable_workspace_fields.rs` better?

File: src/cargo/util/toml/mod.rs
Comment (epage) [reply src/cargo/util/toml/mod.rs:1039]: I've noted this in the tracking issue.  marking it as resolved.

File: tests/testsuite/deduplicate_workspace.rs
Comment (epage) [reply tests/testsuite/deduplicate_workspace.rs:1]: That works.

File: tests/testsuite/deduplicate_workspace.rs
Comment (Muscraft) [reply tests/testsuite/deduplicate_workspace.rs:130]: I personally think `version.workspace = true` looks better and is more descriptive on fields. I can go either way on this and will do whatever you recommend. I would have to change all instances of `{ workspace = true }` to `version.workspace = true` in comments and tests but this isn't a problem. I think we should leave it as `{ workspace = true }` for dependencies since they can also have `features` and `optional`.

I can make this change right now as I still have to push a commit for other fixes.

File: tests/testsuite/deduplicate_workspace.rs
Comment (Muscraft) [reply tests/testsuite/deduplicate_workspace.rs:498]: I'll change it to this!

File: tests/testsuite/deduplicate_workspace.rs
Comment (Muscraft) [reply tests/testsuite/deduplicate_workspace.rs:501]: I'll change it to this!

File: tests/testsuite/deduplicate_workspace.rs
Comment (Muscraft) [reply tests/testsuite/deduplicate_workspace.rs:727]: I'll change it to this!

File: tests/testsuite/deduplicate_workspace.rs
Comment (epage) [reply tests/testsuite/deduplicate_workspace.rs:130]: I can go either way on dependencies.  The same could be said for `dep = "version"` but I also dislike having to change the structure of data to insert a field (similarly, I personally dislike `use` statements and especially `use {};`, thats just asking for merge conflicts)

**Comment by Muscraft** [comment]
>Looking over the RFC, a couple more items I've noticed that we don't have covered
> - Should we have a test to verify cargo metadata and cargo read-manifests behavior?
> - Looks like we should add to our task list the Cargo.lock handling as I don't think I saw it in this PR

Would these not be covered by other tests in cargo? or are you meaning explicit tests that have `{ workspace = true }`? For the most part, I think that these two things shouldn't be an issue since as we `resolve` things we overwrite the original `TomlManifest` with the resolved values. `Cargo.lock` showing warnings for unused dependencies is the only thing I could see going on but I'm not sure how to test for that or metadata. 

>> Changing to a impl FnOnce() -> CargoResult<T> did have some benefits and made things less cluttered overall, but it made the getters for InheritableFields a little verbose. If the getters need to be fixed, we could do a macro_rules! but I dislike this for readability and maintenance reasons. If you have any suggestions I think it might be a pain point.

>I think it looks great. Putting it on the inheritable fields helps keep all the error reporting in one place so we can keep them in sync.

The only way I think we could keep them more in sync would be to make a function in `InheritbaleFields` to handle the error message. Something like 
```rust
fn error(&self, field: &str) -> anyhow::Error {
    anyhow!("`workspace.{}` was not defined", field)
}
```
then each getter would be similar to:
```rust
pub fn dependencies(&self) -> CargoResult<BTreeMap<String, TomlDependency>> {
    self.dependencies.clone().map_or(
        Err(self.error("dependencies")),
        |d| Ok(d),
    )
}
```
This is nitpicky and doesn't have to be done. The real thing would be to get rid of the very similar 
```rust
map_or(Err(anyhow!("message")), |i|  Ok(i)))
```
I tried to get rid of this but I made things a bit odd in my brief attempt.

**Comment by epage** [comment]
> Would these not be covered by other tests in cargo? or are you meaning explicit tests that have `{ workspace = true }`? For the most part, I think that these two things shouldn't be an issue since as we `resolve` things we overwrite the original `TomlManifest` with the resolved values. `Cargo.lock` showing warnings for unused dependencies is the only thing I could see going on but I'm not sure how to test for that or metadata.

Yes, I was thinking of when they are using `{workspace = true}`.  I would lean towards explicitly having test to make sure our user-facing requirements are tested and will ensure they stay tested through refactors but we can also see what others from the cargo team say.

> The only way I think we could keep them more in sync would be to make a function in `InheritbaleFields` to handle the error message. Something like

If consolidating the error message makes the code worse, I'm personally fine having it in each function.  Having them all in one tight location where its obvious that you should update others when updating one is a big help!

File: tests/testsuite/deduplicate_workspace.rs
Comment (Muscraft) [reply tests/testsuite/deduplicate_workspace.rs:130]: I think this is something we should add to a cleanup/optimization PR. Changing all `field = { workspace = true }` to `field.workspace = true` might be outside of the scope of this PR and is something that we should talk about in the RFC.

Comment (epage) [reply tests/testsuite/deduplicate_workspace.rs:130]: I've put it in the tracking issue regarding documentation

Marking this as resolved.

**Comment by Muscraft** [comment]
> Yes, I was thinking of when they are using {workspace = true}. I would lean towards explicitly having test to make sure our user-facing requirements are tested and will ensure they stay tested through refactors but we can also see what others from the cargo team say.

Sounds good, I saw you added it to the RFC so I will add the tests for it then.

**Review by epage — APPROVED** [review]
Thanks for all the work on this!

**Comment by Muscraft** [comment]
> Thanks for all the work on this!

Anytime! It's been something I've been meaning to work on for a while and got time to do so. I will start work on Part 2 once this gets merged but it may take me a bit to get reacclimated with what I need to do.

**Comment by epage** [comment]
Going ahead with it as its feature gated, well tested, and should have minimal impact for existing users.

@bors r+

**Pull request merged**
